crates/oci-distribution/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+#[macro_use]
+extern crate serde;
+
+use hyperx::header::Header;
+use oauth2::{Client as OauthClient, StandardToken};
+use serde::Deserialize;
+use std::convert::TryFrom;
+use url::Url;
+use www_authenticate::{Challenge, ChallengeFields, RawChallenge, WwwAuthenticate};
+
+use crate::errors::*;
+use crate::reference::Reference;
+
+const OCI_VERSION_KEY: &str = "Docker-Distribution-Api-Version";
+
+pub mod errors;
+pub mod reference;
+
+type OciResult<T> = Result<T, anyhow::Error>;
+
+struct Client {
+    token: Option<StandardToken>,
```

Comment:
To prevent users from using an unauthenticated Client, I would make this not optional and have a constructor that does the authencating and populates this field.

Reply:
Clients are supposed to have the choice of authenticating (because some registries don't support an oauth flow). I made this clearer in the inline docs